### ISSUE TITLE
Search query is entirely broken if one git repo generates a bad archive

### ISSUE THREAD

**Issue by slimsag** [sourcegraph/sourcegraph#3799]
When https://github.com/sourcegraph/sourcegraph/issues/3798 occurred for me the search UI showed this error:

```
1 error occurred: * text search failed: failed to search 50.112.230.151/TEST/ssw-scientific--nndd@: failed to fetch missing archive cache item: failed to copy and close missing archive cache item: archive/tar: invalid tar header
```

What is surprising however is that despite the fact that the error only affects that one repository, it breaks the entire search and I get zero results:

```
type:file err
```

![image](https://user-images.githubusercontent.com/3173176/57107415-c67dfe80-6ce4-11e9-83b5-4eeab1447681.png)

But omitting that repository from my query does give me results:

```
type:file err -repo:TEST\/ssw-scientific
```

![image](https://user-images.githubusercontent.com/3173176/57107454-e9a8ae00-6ce4-11e9-93a3-094ab4cfcc1e.png)

This is a surprising bug because we have code and a UX in place to explicitly make it such that one bad repo does not ruin the entire search already. So the question is, why doesn't it work in this case?

To reproduce, you may either follow reproduction steps in #3798 (hard) OR simply change this method to always return an error for one specific repository on your dev instance: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@95fb2bc/-/blob/cmd/searcher/main.go#L54

cc @sourcegraph/code-search

**Comment by sqs** [comment]
@ijt @ijsnow This seems important enough to merit being in 3.4. At this point I know there isn't much feature stuff to remove to make time, so it's basically a bug queue and addition of tests. Overall this is lower in priority relative to the other bugs you're currently working on because it is rare. (If I am wrong about this, let me know. I haven't seen or heard of this bug occurring before.)

**Comment by slimsag** [comment]
> because it is rare. (If I am wrong about this, let me know. I haven't seen or heard of this bug occurring before.)

Agreed. I'll report back here if I see it anywhere else or in a customer deployment.

**Comment by ijt** [comment]
I tried it with fakehub by changing one of the repo entries in the config to have `-TYPO` at the end, but it didn't trigger the issue. I'll try again with the `searcher` hack.

**Comment by ijt** [comment]
I also tried the searcher hack but it didn't repro for me.

**Comment by ijt** [comment]
@slimsag can you verify that this is still an issue?

**Comment by ijsnow** [comment]
Same trouble here https://github.com/sourcegraph/sourcegraph/issues/3798#issuecomment-489813562

**Comment by ijt** [comment]
Moving this off of 3.4 due to no repro.

**Comment by slimsag** [comment]
I saw this on a customer instance (the one I manage), and on my replica of that instance a 2nd time.

Also, it appears you reproduced this @ijt in https://github.com/sourcegraph/issues-uber/issues/179#issuecomment-490353387 so we can use those same steps to reproduce this I think.

**Comment by slimsag** [comment]
@ggilmore also just saw this on k8s.sgdev.org

**Comment by slimsag** [comment]
Given when I reported this, I think this could be a regression in 3.3

**Comment by slimsag** [comment]
There are four issues here here:

1. When the frontend cannot connect to `searcher`, e.g. due to it being overloaded or the docker/k8s network failing in some way, this same type of issue occurs. #4101 is tracking one such instance of this where k8s.sgdev.org's frontend cannot correctly communicate with `searcher` and it is unclear why.

2. [ ] When the code host is unreachable and repositories are not clonable (which is what this issue is tracking). It is unclear why @ijt couldn't reproduce this above, maybe he had to wait until the repository updated OR maybe this is because repository updating doesn't seem to work with fakehub which he was using to test this.

3. The UI reports this in multiple different poor ways (random timeout errors from search are seen when this occurs, effectively -- @ijt saw this when trying to reproduce https://github.com/sourcegraph/issues-uber/issues/179#issuecomment-490353387 -- there is no tracking issue for this and it is probably not worth tackling.

4. [ ] The fact that when such an error occurs, we outright refuse to return any search results instead of just erroring out on that specific subset of results (i.e. we don't provide the user with the index search results which would still work fine as indicated in this issue description). This is being tracked in this issue.

**Comment by keegancsmith** [comment]
I'll tackle this issue (and whatever else is mentioned in the comments) tomorrow.

**Comment by ggilmore** [comment]
@keegancsmith see also https://github.com/sourcegraph/sourcegraph/issues/3798

**Comment by keegancsmith** [comment]
> 1 error occurred: * text search failed: failed to search 50.112.230.151/TEST/ssw-scientific--nndd@: failed to fetch missing archive cache item: failed to copy and close missing archive cache item: archive/tar: invalid tar header

This is a very surprising error message since it indicates two things:
- We are asking searcher to search for something without a commit (notice after the `@` there is nothing)
- We only send a request to searcher once we have resolved a commit => the repository is cloned. However, this error message indicates it is coming from searcher.
- Another possibility: I see the repository name is uppercase. We could have a case sensitivity issue somewhere.

I'm trying to reproduce but can't right now. Is it possible we added caching somewhere for resolving the head commit, and that is misbehaving / breaking the invariant that searcher assumes it gets a resolved commit that exists on gitserver.

**Comment by keegancsmith** [comment]
Is there a way to reproduce this? Any of our environments exposing this behaviour. I'm a bit confused how to proceed further, since I haven't been able to reproduce and from reading the code I am unsure how this is possible other than rare networking issues.

Either way I'll send out a patch which "papers" over this issue and treats the failure as ephemeral in the graphql layer.

**Comment by ggilmore** [comment]
@keegancsmith Try this link: 

https://k8s.sgdev.org/search?q=lang:go

![Screen Shot 2019-05-23 at 11 50 38 AM](https://user-images.githubusercontent.com/9022011/58278519-120a5200-7d51-11e9-8527-9bcf256fe902.png)

https://app.lightstep.com/sourcegraph-dev/trace?span_guid=a75981819b237a0&at_micros=1558637486641918

**Comment by slimsag** [comment]
Reopening because it sounds like we can deploy #4184 to k8s.sgdev.org and confirm if it works there

**Comment by keegancsmith** [comment]
Workaround seems to be working on k8s.sgdev.org

It seems git is generating bad archives for this repo. It isn't having any trouble cloning / etc. 

```
$ kubectl exec gitserver-3 -- git --git-dir /data/repos/gitolite.sgdev.org/13k/SSW-SCIENTIFIC/NNDD/.git archive --worktree-attributes --format=tar c21b98da2ca7f007230e696b2eda5da6589fe137 -- > nndd.tar
$ tar tf nndd.tar > /dev/null
tar: Ignoring malformed pax extended attribute
tar: Error exit delayed from previous errors.
```

I tried to dive in a bit more to work out how to get git to generate a better archive. Also explored working around it in go's tar/archive reader. Both were dead ends. I've reported this bug to git: https://public-inbox.org/git/CAMVcy0Q0TL6uEGR2NeudJrOiXdQ87XcducL0EwMidWucjk5XYw@mail.gmail.com/T/#u

In the meantime this workaround will have to do until git is patched.

**Comment by slimsag** [comment]
Only thing left here is to cherry-pick this onto 3.4 (and maybe some other changes, not sure what else is important recently) and do another patch release.

**Comment by slimsag** [comment]
Indeed, I can confirm that when I was seeing this originally when I reported this issue & reproducing it now that it is limited to just that one repository it seems.

**Comment by keegancsmith** [comment]
Cherry-picked onto 3.4 as 9b2a57b05d

**Comment by keegancsmith** [comment]
Reported the issue to golang as well. It seems `archive/tar`'s goal is to be able to read any archive gnutar or bsdtar can read. So this is an upstream issue in golang. The fix is easy, but may lead to other weird issues. https://github.com/golang/go/issues/32264

Either way I believe this issue should be considered resolved. Fix in master and 3.4 release branch.
